### PR TITLE
zihintntl Intrinsic function proposal

### PR DESCRIPTION
**Pull request by BeMg**
This is a proposal for zihintntl intrinsic function.

### CODE REVIEW

**Comment by asb** [comment]
For further context, I understand this intends to match what Clang currently provides (which is gated on enabling experimental extension support right now).

**Comment by kito-cheng** [comment]
For more context: this also match what @cmuellner proposed https://github.com/riscv-non-isa/riscv-c-api-doc/issues/30#issuecomment-1200199130 before

**Comment by kito-cheng** [comment]
c.c. @aswaterman

**Comment by aswaterman** [comment]
LGTM.  The only comment I'd make is that we'd consider adding overloaded versions that omit `domain`, implying `ALL`.  But I don't know offhand whether overloaded intrinsics are valid in C, as opposed to C++, and maybe this "easy-mode" will prove controversial for other reasons--so if there's disagreement, I don't want to block forward progress.

**Comment by nick-knight** [comment]
> But I don't know offhand whether overloaded intrinsics are valid in C (as opposed to C++),

Both GCC's and Clang's C-language dialects support a limited form of C++ function overloading. We use this already in the vector intrinsics ("overloaded API"). Since these are the only compilers we require to prove concepts, this seems actionable here as well.

**Comment by BeMg** [comment]
> LGTM. The only comment I'd make is that we might consider adding overloaded versions that omit the `domain` argument, implying `ALL`. But I don't know offhand whether overloaded intrinsics are valid in C (as opposed to C++), and maybe this "easy-mode" will prove controversial for other reasons--so if there's disagreement, I don't want to block forward progress.

The version that omit the `domain` argument I think the clang exists the similar built-in function `__builtin_nontemporal_load/store` [1]. 

[1] https://clang.llvm.org/docs/LanguageExtensions.html#non-temporal-load-store-builtins

**Comment by aswaterman** [comment]
@BeMg OK.  Can we document that API here, including that it implies `ALL`, so that GCC and LLVM will match?  Or is the problem that we can't make it part of this API because it does not include the `__riscv` prefix?

**Comment by BeMg** [comment]
I'm not sure whether to add the `__builtin_nontemporal_load/store` to proposal. Besides it doesn't not including the `__riscv` prefix, GCC seem doesn't has these built-in functions now.

cc @kito-cheng 

The overloaded version look like more consistent, so I add it into proposal. 

It could be implemented by some macros. 

```
#define __riscv_ntl_load_with_domain(PTR, DOMAIN) __builtin_riscv_ntl_load((PTR), (DOMAIN))
#define __riscv_ntl_load_without_domain(PTR) __builtin_riscv_ntl_load((PTR), __RISCV_NTLH_ALL)

#define SELECT_NTL_LOAD(_1,_2,NAME,...) NAME
#define __riscv_ntl_load(...) SELECT_NTL_LOAD(__VA_ARGS__, __riscv_ntl_load_with_domain, __riscv_ntl_load_without_domain)(__VA_ARGS__)
```

**Review by kito-cheng — APPROVED** [review]
LGTM, seems like the overloading trick is work both on GCC and clang/LLVM, and I will try to find some resource to implement GCC part.

I would like to get explicitly ack from @aswaterman and @asb before merge.

**Comment by aswaterman** [comment]
OK with me, too.

**Comment by BeMg** [comment]
I will implement the overloaded version in LLVM.

**Comment by kito-cheng** [comment]
Thanks for everyone, I gonna merge this :)

File: riscv-c-api.md [hunk 1]
```diff
@@ -235,6 +235,51 @@ long __riscv_clmul (long a, long b); // clmul rd, rs1, rs2
 vint8m1_t __riscv_vadd_vv_i8m1(vint8m1_t vs2, vint8m1_t vs1, size_t vl); // vadd.vv vd, vs2, vs1
 ```
 
+### NTLH Intrisics 
+
+The RISC-V zihintntl extension provides the RISC-V specific intrinsic functions for generating non-temporal memory accesses. These intrinsic functions provide the domain parameter to specify the behavior of memory accesses.
```
Comment (topperc) [inline riscv-c-api.md:240]: zihintntl -> Zihintntl

File: riscv-c-api.md [hunk 2]
```diff
@@ -235,6 +235,51 @@ long __riscv_clmul (long a, long b); // clmul rd, rs1, rs2
...
+
+```
+type __riscv_ntl_load (type *ptr, int domain);
+void __riscv_ntl_store (type *ptr, type val, int domain);
+```
+
+There are overloaded functions of `__riscv_ntl_load` and `__riscv_ntl_store`. When these intrinsic functions omit the `domain` argument, the `domain` is implied as `__RISCV_NTLH_ALL`.
+
+```
+type __riscv_ntl_load (type *ptr);
```
Comment (topperc) [inline riscv-c-api.md:255]: Is this actually implemented? I don't see a test for it in clang

**Pull request merged**